### DISCUSSION TITLE
Hardware Library API

### DISCUSSION THREAD

**Discussion by MeisterBob** [zigpy/zigpy#709] [ZHA development]
I started developing a radio library for nRF52 SoCs but I am a little Stuck, because I don't really understand the API between zigpy and the radio libraries. Can someone (maybe a developer of the other radio libraries) share his knowledge?

**Comment by puddly** [comment]
A radio library is just a Python package that exposes `your_radio_lib.zigbee.application.ControllerApplication`, a subclass of [`zigpy.application.ControllerApplication`](https://github.com/zigpy/zigpy/blob/dev/zigpy/application.py).

Your radio library will be created with `app.config` set to the zigpy config schema:

```Python
from your_radio_lib.zigbee.application import ControllerApplication

app = await ControllerApplication.new({
    "device": {
        "path": "/dev/serial/by-id/..."
    }
}, auto_form=True, start_radio=True)
```

Minimally, you need to implement abstract methods in the base `ControllerApplication` class:

 - `probe` classmethod: called by ZHA to determine if the config for a serial port (i.e. `{"device": "/dev/serial/by-id/..."}`) can be handled by your radio library. You can just connect to the port, send some no-op command to ensure it's for a nRF52 SoC, and clean up.
 - `startup`: connects to the serial port and starts the Zigbee radio. `auto_form=True` is usually passed, so you should be able to call `startup` on an application instance pointing to either an unconfigured radio or to an already-configured radio and both should end up in the same state.
 - `form_network`: forms a network with the passed config dictionary. If a value is not provided (for example, a network key, an extended PAN ID, or a PAN ID), you need to randomly generate it.
 - `shutdown`: disconnect from the serial port and clean everything up.
 - `request`: send a unicast request to device. Some radios need to have endpoints explicitly registered (e.g. TI) so you can re-write `src_ep` to the appropriate endpoint for the request's profile and cluster.
 - `permit_ncp`: open the network for joins. Some radios properly handle loopback requests (i.e. sending a "permit join" request to `0x0000` from` 0x0000`) so this may not be necessary.

Received messages should be passed to `handle_message`.

This will get your radio library functioning enough to use. More methods should be implemented for full functionality:
 - `update_network`: you can ignore this, it's unused.
 - `force_remove(device)`: removes a device from the network if the device itself will not leave. If your radio doesn't expose anything like this, make this method a no-op.
 - `mrequest`: send a group request.
 - `broadcast`: send a broadcast request.
 - `permit_with_key`: permit a device with a specific IEEE and install code to join.

Some attributes need to be set for metadata to show up in ZHA:

 - `_channel`: the network channel
 - `_channels`: the radio channel mask (not used)
 - `_ext_pan_id`: the network extended PAN ID
 - `_ieee`: the coordinator's IEEE address
 - `_nwk`: the coordinator's NWK address (always 0x0000)
 - `_nwk_update_id`: the network update ID, if this is exposed by your radio library.
 - `_pan_id`: the network PAN ID
